pkg/splunk/enterprise/afwscheduler.go

```diff
@@ -0,0 +1,582 @@
+	afwPipeline.phaseWaiter.Add(1)
+	go afwPipeline.installPhaseManager(&afwPipeline.phaseWaiter, afwPipeline.sigTerm)
+
+	scopedLog.Info("Creating pipeline workers for pending app packages")
+	for appSrcName, appSrcDeployInfo := range appDeployContext.AppsSrcDeployStatus {
+		for _, deployInfo := range appSrcDeployInfo.AppDeploymentInfoList {
+			var pplnPhase enterpriseApi.AppPhaseType
+			var podName string
+			var podID int
+
+			appSrc, err := getAppSrcSpec(appFrameworkConfig.AppSources, appSrcName)
+			if err != nil {
+				// Error, should never happen
+				continue
+			}
+
+			// Track the cluster scoped apps to track the bundle push
+			if appSrc.Scope == enterpriseApi.ScopeCluster && deployInfo.DeployStatus != enterpriseApi.DeployStatusComplete {
+				clusterScopedApps = append(clusterScopedApps, &deployInfo)
+			}
+
+			pplnPhase = ""
+			//Push All the Intermediatory work to the Pipeline phases and let the corresponding phase manager take care of them
+			if deployInfo.PhaseInfo.Phase == enterpriseApi.PhaseDownload && deployInfo.PhaseInfo.RetryCount < pipelinePhaseMaxRetryCount {
+				pplnPhase = enterpriseApi.PhaseDownload
+			} else if deployInfo.PhaseInfo.Phase == enterpriseApi.PhasePodCopy && deployInfo.PhaseInfo.RetryCount < pipelinePhaseMaxRetryCount {
+				pplnPhase = enterpriseApi.PhasePodCopy
+			} else if deployInfo.PhaseInfo.Phase == enterpriseApi.PhaseInstall && deployInfo.PhaseInfo.Status != enterpriseApi.AppPkgInstallComplete && deployInfo.PhaseInfo.RetryCount < pipelinePhaseMaxRetryCount {
+				// Cluster scopes do not get to this phase. Only local scoped apps fall into this pipeline.
+				// For cluster scoped apps, all that we need is bundle push. So, a one-shot worker(outside of this logic) serves the purpose
+				pplnPhase = enterpriseApi.PhaseInstall
+			}
+
+			// Ignore any other apps that are not in progress
+			if pplnPhase != "" {
+				podName = getApplicablePodNameForWorker(cr, podID)
+				sts := afwGetReleventStatefulsetByKind(cr, client)
+				afwPipeline.createAndAddPipelineWorker(pplnPhase, &deployInfo, appSrcName, podName, appFrameworkConfig, client, cr, sts)
+			}
+		}
+	}
+
+	var clusterAppsList string
+	for _, appDeployInfo := range clusterScopedApps {
+		clusterAppsList = fmt.Sprintf("%s:%s, %s", appDeployInfo.AppName, appDeployInfo.ObjectHash, clusterAppsList)
+	}
+	scopedLog.Info("List of cluster scoped apps(appName:digest) for this reconcile entry", "apps", clusterAppsList)
+
+	// Wait for the yield function to finish
+	afwPipeline.phaseWaiter.Add(1)
+	go func() {
+		afwEntryTime := time.Now().Unix()
+		for {
+			if afwEntryTime+yieldTime < time.Now().Unix() || afwPipeline.isPipelineEmpty() {
+				scopedLog.Info("Yielding from AFW scheduler", "time elapsed", time.Now().Unix()-afwEntryTime)
+				// terminate download Phase
+				afwPipeline.sigTerm <- true
+
+				// terminate podCopy Phase
+				afwPipeline.sigTerm <- true
+
+				// terminate install Phase
+				afwPipeline.sigTerm <- true
```

Comment:
I think instead of this you can just close the `sigTerm` channel and check the return bool value when you read the channel in other places, which is indication of a closed channel. What do you think?

Reply:
I started with a select on the receiving side and ended up this way. But, yes, sending a value is not doing anything additional here, and a simple channel closer will be cleaner.